### PR TITLE
vsr: add concept of fatal errors

### PR DESCRIPTION
**Pull request by matklad**
Use assert to spot bugs, use `fatal` to crash intentionally when operating further is impossible.

### CODE REVIEW

File: src/vsr.zig [hunk 1]
```diff
@@ -638,6 +638,35 @@ pub const UpgradeRequest = extern struct {
     }
 };
 
+/// To ease investigation of accidents, assign a separate exist status for each fatal condition.
```
Comment (cb22) [inline src/vsr.zig:641]: :heart:

File: src/lsm/node_pool.zig [hunk 2]
```diff
@@ -60,7 +62,12 @@ pub fn NodePoolType(comptime _node_size: u32, comptime _node_alignment: u13) typ
         pub fn acquire(pool: *NodePool) Node {
             // TODO: To ensure this "unreachable" is never reached, the primary must reject
             // new requests when storage space is too low to fulfill them.
-            const node_index = pool.free.findFirstSet() orelse unreachable;
+            const node_index = pool.free.findFirstSet() orelse vsr.fatal(
+                .manifest_node_pool_exhausted,
+                \\out of memory for manifest, restart the replica with larger '--memory-lsm-manifest' argument
+            ,
+                .{},
+            );
```
Comment (cb22) [inline src/lsm/node_pool.zig]: Is there perhaps a better way to wrap this? The `,` on a single line is throwing me off!

File: src/vsr/replica.zig [hunk 3]
```diff
@@ -7887,6 +7887,7 @@ pub fn ReplicaType(
                 .updating_checkpoint => |checkpoint| {
                     if (checkpoint.header.op == self.op_checkpoint()) {
                         self.sync_superblock_update_finish();
+                        assert(self.syncing == .idle);
```
Comment (cb22) [inline src/vsr/replica.zig]: Drive-by assert, or snuck in? (and below)

File: src/lsm/node_pool.zig (hunk 2 above)
Comment (matklad) [reply src/lsm/node_pool.zig]: Yeah, found a better way, thanks!

File: src/vsr/replica.zig (hunk 3 above)
Comment (matklad) [reply src/vsr/replica.zig]: Oups, apparently I can't git today, fixed!

**Review by cb22 — APPROVED** [review]

File: src/vsr.zig (hunk 1 above)
Comment (vegarsti) [reply src/vsr.zig:641]: Typo in `exist status` but I won't bother you with a PR for such a minor change!

Comment (matklad) [reply src/vsr.zig:641]: Thanks, fixed!
